File: src/main/java/io/github/pavelshe11/messengermicro/store/entities/MessageEntity.java
package io.github.pavelshe11.messengermicro.store.entities;

import io.github.pavelshe11.messengermicro.store.enums.MessageStatusType;
import jakarta.persistence.*;
import lombok.*;

import java.util.List;
import java.util.UUID;

@Setter
@Getter
@NoArgsConstructor
@AllArgsConstructor
@Builder
@Entity
@Table(name = "message")
public class MessageEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @Column(nullable = false, updatable = false)
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "chat_id", referencedColumnName = "id", nullable = false)
    private ChatRoomEntity chatRoom;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "sender_id", referencedColumnName = "id", nullable = false)
    private ChatSendersEntity chatSenders;

    @Column(name = "message_text")
    private String messageText;

    @Column(name = "status", nullable = false)
    @Enumerated(EnumType.STRING)
    private MessageStatusType messageStatusType;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "parent_message_id")
    private MessageEntity parentMessage;

    @OneToMany(mappedBy = "parentMessage", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<MessageEntity> replies;
}

File: src/main/java/io/github/pavelshe11/messengermicro/store/entities/MessageEntity.kt
package io.github.pavelshe11.messengermicro.store.entities

import io.github.pavelshe11.messengermicro.store.enums.MessageStatusType
import jakarta.persistence.*
import java.time.Instant
import java.util.*

@Entity
@Table(name = "message")
data class MessageEntity (
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @Column(nullable = false, updatable = false)
    val id: UUID? = null,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "chat_id", referencedColumnName = "id", nullable = false)
     var chatRoom: ChatRoomEntity,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "sender_id", referencedColumnName = "id", nullable = false)
     var chatSenders: ChatSendersEntity,

    @Column(name = "message_text", nullable = false)
     var messageText: String,

    @Column(name = "status", nullable = false)
    @Enumerated(EnumType.STRING)
     var messageStatusType: MessageStatusType,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "parent_message_id")
    var parentMessage: MessageEntity? = null,

    @Column(nullable = false)
    var draft: Boolean = false,

    @Column(name = "sending_time", nullable = false)
    var sendingTime: Instant = Instant.now(),

    @Column(name = "updated_at")
    var updatedAt: Instant = Instant.now(),

    @OneToMany(mappedBy = "parentMessage", cascade = [CascadeType.ALL], orphanRemoval = true)
     var replies: List<MessageEntity> = emptyList()
)
